### ISSUE TITLE
Question about the implementation of the IPO

### ISSUE THREAD

**Issue by aygonl** [PKU-MARL/Safe-Policy-Optimization#18]
I am confused about the implementation of the IPO. The original paper uses the logarithm barrier function to approximate the indicator function. But 'log' does not appear in this [code](https://github.com/PKU-MARL/Safe-Policy-Optimization/blob/main/safepo/algos/ipo.py)

**Comment by JuntaoDai** [comment]
In IPO, the loss function is 
![image](https://user-images.githubusercontent.com/93032279/196993928-2ff8d83f-0432-43e6-b204-f57974f62bf8.png)
whose gradient is 
![image](https://user-images.githubusercontent.com/93032279/196994159-6e53aed0-17b0-4380-860b-2c19c5027ceb.png)
Thus, we rewrite the loss function as follows, 
![image](https://user-images.githubusercontent.com/93032279/196994582-949d6353-c0d6-4988-a9aa-b23012aa55f3.png)
to avoid the ill-conditioned issue, where $\lambda_{i,\theta}=\min(\lambda_i(\theta), \lambda_\text{max})$ is a parameter related to $\theta$ but detached when calculating the gradient flow. Notably, $\mathcal L^\lambda(\theta)$ and $\mathcal L(\theta)$ have the same gradient when $\lambda_{i,\theta}=\lambda_i(\theta)$.

The idea of $\mathcal L^\lambda(\theta)$ is to truncate $\lambda_i(\theta)$ at a sufficiently large number $\lambda_\text{max}$ to both ensure the suppression effect and avoid the pathological curvature problem in the gradient descent caused by the Hessian matrix with a large conditional number.
